[ADD] subscriptions: add scheduled actions doc (#7663)
odoo/documentation: content/applications/sales/subscriptions/scheduled_actions.rst

=================
Scheduled actions
=================

> In Odoo *Subscriptions*, there are two scheduled actions that manage subscriptions. They are turned

Reviewer: Would say _specifically_ what is managed by the scheduled actions
```suggestion
In Odoo *Subscriptions*, there are two scheduled actions that initiate the billing process for active recurring subscriptions, as well as when billing should stop due to subscription expiration. They are turned
```

.. important::
   To access these scheduled actions, *Developer Mode* **must** be activated.

   To activate developer mode, navigate to :menuselection:`Settings app --> Scroll down to Developer
   Tools section --> Activate the developer mode`.

With developer mode activated, navigate to :menuselection:`Settings app --> Technical --> Scheduled
Actions`.

.. image:: scheduled_actions/scheduled-actions-technical-settings-page.png
   :align: center
   :alt: The scheduled actions option under the technical menu in the Odoo Settings application.

Doing so reveals a dedicated :guilabel:`Scheduled Actions` page. On this page, there is a complete
list of scheduled actions for the entire database.

From here, enter `Subscription` in the search bar. Doing so provides three subscription-specific
results. The following documentation focuses on the last two results in the list:

- :guilabel:`Sale Subscription: generate recurring invoices and payments`
- :guilabel:`Sale Subscription: subscriptions expiration`

.. image:: scheduled_actions/scheduled-actions-page-subscription-results.png
   :align: center
   :alt: The subscription-related results on the scheduled actions page in Odoo Settings.

> Ensure the scheduled actions are :guilabel:`Active`. If there is a ticked checkbox in the

Reviewer: We are ensuring this, but why?
The instruction is good and clear, but it's not clear _why_ we're doing this yet (especially if the default in Odoo is that they're active)

Author: Just added a brief sentence above this to explain that if a Scheduled Action is not active, it will not work/the scheduled action will not be performed.

:guilabel:`Active` column for the desired subscription-related scheduled action on the
:guilabel:`Scheduled Actions` page, that means they are `Active`.

If a scheduled action needs to be activated, click into the desired scheduled action from the list.

.. image:: scheduled_actions/scheduled-action-form.png
   :alt: The scheduled action form in the Odoo Settings application.

Then, from the scheduled action form, toggle the switch in the :guilabel:`Active` field to the
right. Doing so turns the switch green, indicating that the scheduled action is now `Active`.

> The ability to set up how often the scheduled action runs is also available on the scheduled action

Reviewer: Might be nice to  include what a scheduled action actually is, and link to any existing documentation on it...

form, in the :guilabel:`Execute Every` field.

.. important::
   The scheduled action will **not** function correctly if the execution time is less than five
   minutes. This is a general rule for all scheduled actions.

   For more information, read the :doc:`Frequent Technical Questions
   </administration/odoo_sh/advanced/frequent_technical_questions>` documentation.

Generate recurring invoices and payments
========================================

In order for the :guilabel:`Sale Subscription: generate recurring invoices and payments` scheduled

> action to run properly, *Deferred Expense* and *Deferred Revenue* accounts **must** set up.

Reviewer: In this language: "in order for the scheduled action to run properly," *properly* is still not quite defined at this point in the documentation; I'd include some definitions earlier for why we're doing what we're doing and what scheduled actions actually are, so when we say *properly* people can understand what that means (especially since they're turned on and configured already by default).

Reviewer: Why also must these be set up? What are they and how do they connect the scheduled action?

To set up *Deferred Expense* and *Deferred Revenue* accounts, navigate to :menuselection:`Accounting
app --> Configuration --> Settings`. Both accounts can be configured in the :guilabel:`Default
Accounts` section.

.. image:: scheduled_actions/deferred-settings-accounting.png
   :align: center
   :alt: The necessary deferred account settings in the Odoo Accounting app's settings page.

> Once the necessary information and configurations are entered in the :guilabel:`Deferred Expense`

Reviewer: what "necessary information and configurations," specifically?

Author: I removed that phrase and replaced it with a more specific mention of the correct accounts needing to be entered in those drop-down menu fields.

and :guilabel:`Deferred Revenue` fields, click :guilabel:`Save` in the upper-left corner.

Create invoice
--------------

The main function of the :guilabel:`Sale Subscription: generate recurring invoices and payments`

> scheduled action revolves around generating recurring invoices for subscriptions.

Reviewer: here we have a definition, can we move it up earlier?

Elements related to this scheduled action can be found on confirmed subscription sales orders.

To examine these elements, open any confirmed sales order in the *Subscriptions* application to
reveal the subscription sales order form.

On a confirmed subscription sales order form, focus on the :guilabel:`Recurring Plan` and
:guilabel:`Date of Next Invoice` fields.

.. image:: scheduled_actions/confirmed-subscription-sales-order-fields.png
   :align: center
   :alt: A confirmed subscription sales order in the Odoo Subscriptions application.

The scheduled action creates an invoice when today's date is the same date as the :guilabel:`Date of
Next Invoice`.

Odoo uses the information in the :guilabel:`Recurring Plan` field to update the next invoice date
accordingly.

> .. warning::
>    If there is no payment provider set up, Odoo will **not** create an invoice. Instead, in the
>    *chatter* of the subscription sales order, the following message appears: `Automatic renewal
>    succeeded. Free subscription. Next invoice: [date]. No email sent.`

Reviewer: More context, if this is accurate:

```suggestion
.. warning::
   If there is no payment provider set up, Odoo will **not** create an invoice and the customer will
   not be charged. Instead, the subscription will be processed as a free recurring product and will
   be reflected as such in the *chatter* of the subscription sales order. The following message
   appears: `Automatic renewal succeeded. Free subscription. Next invoice: [date]. No email sent.`
```

Once the invoice for the subscription sales order is created, the invoice can be viewed by clicking
the :guilabel:`Invoices` smart button that appears at the top of the subscription sales order.

An email is sent to the customer notifying them of the recurring subscription charge, *if* there is
a :guilabel:`Payment Token` on the account.

To check if there is a :guilabel:`Payment Token`, open the :guilabel:`Other Info` tab, and look at
the :guilabel:`Payment Token` field, under the :guilabel:`Subscription` section.

.. image:: scheduled_actions/payment-token-field.png
   :align: center
   :alt: The Payment Token field under the Other Info tab on a subscription sales order form.

If there is no :guilabel:`Payment Token` (i.e. payment provider is `Bank Transfer`), the invoice is
created and sent to the customer. The payment **must** be registered manually in this case.

Closing invoices
----------------

The :guilabel:`Sale Subscription: generate recurring invoices and payments` scheduled action also

> has the ability to close a subscription, if the following occurs:

Reviewer: nitpick since this is an array of conditional logic 😛 

```suggestion
has the ability to close a subscription, if the following conditions are met:
```

> - If the subscription has a :guilabel:`Payment Token`, try to charge.
> - If the charge is successful, create and post the invoice.
> - If the charge fails, send reminders periodically.
> - Close the subscription if it continues to fail for more than fourteen days.

Reviewer: how many levels deep can we indent this, so the logic statements match their parents?
would this work?
```suggestion
- If the subscription has a :guilabel:`Payment Token`, try to charge.
    - If the charge is successful, create and post the invoice.
    - If the charge fails, send reminders periodically.
        - Close the subscription if it continues to fail for more than fourteen days.
```

> 
> - If the subscription has no :guilabel:`Payment Token`, create and post the invoice.

Reviewer: would include this as the second parent condition in the list above
```suggestion
- If the subscription has no :guilabel:`Payment Token`, create and post the invoice.
```

Subscriptions expiration
========================

The :guilabel:`Sale Subscription: subscriptions expiration` scheduled action checks for all other

> conditions that may cause a subscription to close automatically.

Reviewer: Does it only check for conditions or does it check for conditions and then close the subscription if certain conditions are met.

Usually a check has a series of outcomes after the check. Does the AA handle the outcomes too?

Author: Added clarification that the scheduled action does, indeed, close the subscription is certain conditions are met.

that particular scheduled action.

Scheduled actions help users save time, accomplish goals, and complete tasks in a quick, efficient
fashion.

In Odoo *Subscriptions*, there are two scheduled actions that initiate the billing process for
active recurring subscriptions, as well as when billing should stop due to subscription expiration.

They are turned on, by default, so keep that in mind while using the *Subscriptions* application.
They can be deactivated at any time, if needed, to manage subscriptions manually.

Access scheduled actions
========================

.. important::
   To access these scheduled actions, *Developer Mode* **must** be activated.

   To activate developer mode, navigate to :menuselection:`Settings app --> Scroll down to Developer
   Tools section --> Activate the developer mode`.

With developer mode activated, navigate to :menuselection:`Settings app --> Technical --> Scheduled
Actions`.

.. image:: scheduled_actions/scheduled-actions-technical-settings-page.png
   :align: center
   :alt: The scheduled actions option under the technical menu in the Odoo Settings application.

> Doing so reveals a dedicated :guilabel:`Scheduled Actions` page. On this page, there is a complete

Reviewer: nitpick 😉 
```suggestion
Doing so reveals a dedicated :guilabel:`Scheduled Actions` dashboard. On this page, there is a complete
```

list of scheduled actions for the entire database.

From here, enter `Subscription` in the search bar. Doing so provides three subscription-specific
results. The following documentation focuses on the last two results in the list:

- :guilabel:`Sale Subscription: generate recurring invoices and payments`
- :guilabel:`Sale Subscription: subscriptions expiration`

.. image:: scheduled_actions/scheduled-actions-page-subscription-results.png
   :align: center
   :alt: The subscription-related results on the scheduled actions page in Odoo Settings.

> If a scheduled action is **not** `Active`, the action does **not** work, and the desired scheduled
> action is **not** performed.
> 
> That is why it is important to ensure the scheduled actions are :guilabel:`Active`. If there is a
> ticked checkbox in the :guilabel:`Active` column for the desired subscription-related scheduled
> action on the :guilabel:`Scheduled Actions` page, that means they are `Active`.
> 
> If a scheduled action needs to be activated, click into the desired scheduled action from the list.

Reviewer: More succinct and orderly
```suggestion
Determine if a scheduled action is active by looking under the :guilabel:`Active` column, in the
it's corresponding row on the :guilabel:`Scheduled Actions` dashboard, for a ticked checkbox; if
the checkbox is green with a check mark, the scheduled action is active.

If a scheduled action needs to be activated, click into the desired scheduled action from the list.
```

.. image:: scheduled_actions/scheduled-action-form.png
   :alt: The scheduled action form in the Odoo Settings application.

Then, from the scheduled action form, toggle the switch in the :guilabel:`Active` field to the
right. Doing so turns the switch green, indicating that the scheduled action is now `Active`.

The ability to set up how often the scheduled action runs is also available on the scheduled action
form, in the :guilabel:`Execute Every` field.

.. important::
   The scheduled action does **not** function correctly if the execution time is less than five
   minutes. This is a general rule for all scheduled actions.

   For more information, read the :doc:`Frequent Technical Questions
   </administration/odoo_sh/advanced/frequent_technical_questions>` documentation.

Generate recurring invoices and payments
========================================

> The main function of the :guilabel:`Sale Subscription: generate recurring invoices and payments`
> scheduled action revolves around generating recurring invoices for subscriptions.

Reviewer: Redundant to the next paragraph, and the label itself. Not needed.
```suggestion
```

In order for the :guilabel:`Sale Subscription: generate recurring invoices and payments` scheduled
action to properly generate recurring invoices and payments on subscriptions, the *Deferred Expense*

> and *Deferred Revenue* accounts **must** set up, in order for Odoo to process various invoices and

Reviewer: Grammar
```suggestion
and *Deferred Revenue* accounts **must** be set up, in order for Odoo to process various invoices and
```

Subscriptions expiration
========================

The :guilabel:`Sale Subscription: subscriptions expiration` scheduled action checks for all other
conditions that may cause a subscription to close automatically. If certain conditions are met, the
scheduled action closes that subscription.

First, the :guilabel:`Sale Subscription: subscriptions expiration` scheduled action checks to see if
the end date has passed, which is configured on the subscription sales order.

.. image:: scheduled_actions/subscription-expiration-date.png
   :align: center
   :alt: The expiration date on a subscription sales order in Odoo Subscriptions.

Then, the :guilabel:`Sale Subscription: subscriptions expiration` scheduled action checks if the
invoice has not been paid within the payment terms deadline.

To access the invoices attached to a subscription, access the sales order for the subscription
product, and click the :guilabel:`Invoices` smart button. Then, look at the :guilabel:`Invoice Date`
column.

.. image:: scheduled_actions/invoices-invoice-date-column.png
   :align: center
   :alt: The Invoice Date column on subscriptions invoice page in Odoo Subscriptions app.

> Unpaid subscriptions with an :guilabel:`Invoice Date` that is past the determined number of days in

Reviewer: Grammar
```suggestion
Unpaid subscriptions with an :guilabel:`Invoice Date` that are past the determined number of days in
```
